File: 16_Bug/java/src/Bug.java
import java.util.ArrayList;
import java.util.Scanner;

/**
 * Game of Bug
 * <p>
 * Based on the Basic game of Bug here
 * https://github.com/coding-horror/basic-computer-games/blob/main/16%20Bug/bug.bas
 * <p>
 * Note:  The idea was to create a version of the 1970's Basic game in Java, without introducing
 * new features - no additional text, error checking, etc has been added.
 */
public class Bug {

    // Dice roll
    public static final int SIX = 6;

    private enum GAME_STATE {
        START,
        PLAYER_TURN,
        COMPUTER_TURN,
        CHECK_FOR_WINNER,
        GAME_OVER
    }

    // Used for keyboard input
    private final Scanner kbScanner;

    // Current game state
    private GAME_STATE gameState;


    private final Insect playersBug;

    private final Insect computersBug;

    // Used to show the result of dice roll.
    private final String[] ROLLS = new String[]{"BODY", "NECK", "HEAD", "FEELERS", "TAIL", "LEGS"};

    public Bug() {

        playersBug = new PlayerBug();
        computersBug = new ComputerBug();

        gameState = GAME_STATE.START;

        // Initialise kb scanner
        kbScanner = new Scanner(System.in);
    }

    /**
     * Main game loop
     */
    public void play() {

        do {
            switch (gameState) {

                // Show an introduction the first time the game is played.
                // And optionally instructions.
                case START:
                    intro();
                    if (!noEntered(displayTextAndGetInput("DO YOU WANT INSTRUCTIONS? "))) {
                        instructions();
                    }

                    gameState = GAME_STATE.PLAYER_TURN;
                    break;

                case PLAYER_TURN:
                    int playersRoll = randomNumber();
                    System.out.println("YOU ROLLED A " + playersRoll + "=" + ROLLS[playersRoll - 1]);
                    switch (playersRoll) {
                        case 1:
                            System.out.println(playersBug.addBody());
                            break;
                        case 2:
                            System.out.println(playersBug.addNeck());
                            break;
                        case 3:
                            System.out.println(playersBug.addHead());
                            break;
                        case 4:
                            System.out.println(playersBug.addFeelers());
                            break;
                        case 5:
                            System.out.println(playersBug.addTail());
                            break;
                        case 6:
                            System.out.println(playersBug.addLeg());
                            break;
                    }

                    gameState = GAME_STATE.COMPUTER_TURN;
                    break;

                case COMPUTER_TURN:
                    int computersRoll = randomNumber();
                    System.out.println("I ROLLED A " + computersRoll + "=" + ROLLS[computersRoll - 1]);
                    switch (computersRoll) {
                        case 1:
                            System.out.println(computersBug.addBody());
                            break;
                        case 2:
                            System.out.println(computersBug.addNeck());
                            break;
                        case 3:
                            System.out.println(computersBug.addHead());
                            break;
                        case 4:
                            System.out.println(computersBug.addFeelers());
                            break;
                        case 5:
                            System.out.println(computersBug.addTail());
                            break;
                        case 6:
                            System.out.println(computersBug.addLeg());
                            break;
                    }

                    gameState = GAME_STATE.CHECK_FOR_WINNER;
                    break;

                case CHECK_FOR_WINNER:
                    boolean gameOver = false;

                    if (playersBug.complete()) {
                        System.out.println("YOUR BUG IS FINISHED.");
                        gameOver = true;
                    } else if (computersBug.complete()) {
                        System.out.println("MY BUG IS FINISHED.");
                        gameOver = true;
                    }

                    if (noEntered(displayTextAndGetInput("DO YOU WANT THE PICTURES? "))) {
                        gameState = GAME_STATE.PLAYER_TURN;
                    } else {
                        System.out.println("*****YOUR BUG*****");
                        System.out.println();
                        draw(playersBug);

                        System.out.println();
                        System.out.println("*****MY BUG*****");
                        System.out.println();
                        draw(computersBug);
                        gameState = GAME_STATE.PLAYER_TURN;
                    }
                    if (gameOver) {
                        System.out.println("I HOPE YOU ENJOYED THE GAME, PLAY IT AGAIN SOON!!");
                        gameState = GAME_STATE.GAME_OVER;
                    }
            }
        } while (gameState != GAME_STATE.GAME_OVER);
    }

    /**
     * Draw the bug (player or computer) based on what has
     * already been added to it.
     *
     * @param bug The bug to be drawn.
     */
    private void draw(Insect bug) {
        ArrayList<String> insectOutput = bug.draw();
        for (String s : insectOutput) {
            System.out.println(s);
        }
    }

    /**
     * Display an intro
     */
    private void intro() {
        System.out.println("BUG");
        System.out.println("CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
        System.out.println();
        System.out.println("THE GAME BUG");
        System.out.println("I HOPE YOU ENJOY THIS GAME.");
    }

    private void instructions() {
        System.out.println("THE OBJECT OF BUG IS TO FINISH YOUR BUG BEFORE I FINISH");
        System.out.println("MINE. EACH NUMBER STANDS FOR A PART OF THE BUG BODY.");
        System.out.println("I WILL ROLL THE DIE FOR YOU, TELL YOU WHAT I ROLLED FOR YOU");
        System.out.println("WHAT THE NUMBER STANDS FOR, AND IF YOU CAN GET THE PART.");
        System.out.println("IF YOU CAN GET THE PART I WILL GIVE IT TO YOU.");
        System.out.println("THE SAME WILL HAPPEN ON MY TURN.");
        System.out.println("IF THERE IS A CHANGE IN EITHER BUG I WILL GIVE YOU THE");
        System.out.println("OPTION OF SEEING THE PICTURES OF THE BUGS.");
        System.out.println("THE NUMBERS STAND FOR PARTS AS FOLLOWS:");
        System.out.println("NUMBER\tPART\tNUMBER OF PART NEEDED");
        System.out.println("1\tBODY\t1");
        System.out.println("2\tNECK\t1");
        System.out.println("3\tHEAD\t1");
        System.out.println("4\tFEELERS\t2");
        System.out.println("5\tTAIL\t1");
        System.out.println("6\tLEGS\t6");
        System.out.println();

    }

    /**
     * Checks whether player entered N or NO to a question.
     *
     * @param text player string from kb
     * @return true if N or NO was entered, otherwise false
     */
    private boolean noEntered(String text) {
        return stringIsAnyValue(text, "N", "NO");
    }

    /**
     * Check whether a string equals one of a variable number of values
     * Useful to check for Y or YES for example
     * Comparison is case insensitive.
     *
     * @param text   source string
     * @param values a range of values to compare against the source string
     * @return true if a comparison was found in one of the variable number of strings passed
     */
    private boolean stringIsAnyValue(String text, String... values) {

        // Cycle through the variable number of values and test each
        for (String val : values) {
            if (text.equalsIgnoreCase(val)) {
                return true;
            }
        }

        // no matches
        return false;
    }

    /*
     * Print a message on the screen, then accept input from Keyboard.
     *
     * @param text message to be displayed on screen.
     * @return what was typed by the player.
     */
    private String displayTextAndGetInput(String text) {
        System.out.print(text);
        return kbScanner.next();
    }

    /**
     * Generate random number
     *
     * @return random number
     */
    private int randomNumber() {
        return (int) (Math.random()
                * (SIX) + 1);
    }
}

File: 16_Bug/ruby/Bug.rb
class BugGame
  YES = ["Y", "y", "Yes", "YES"]
  NO  = ["N", "n", "No", "NO"]

  YOU = { body: 0, neck: 0, head: 0, antennaes: 0, legs: 0, arms: 0, name: "YOU" }
  I   = { body: 0, neck: 0, head: 0, antennaes: 0, legs: 0, arms: 0, name: "I" }
  def initialize
    puts "The Game Bug"
    puts "I HOPE YOU ENJOY THIS GAME."
    instructions
  end

  def run
    loop do
      # YOU FIRST
      play YOU
      if is_completed? YOU
        puts "\n\n\n\nYOU WON"
        break
      end
      puts "\n"

      # I SECOND
      play I
      if is_completed? I
        puts "\n\n\n\nI WON"
        break
      end

      loop do
        puts "Do you want the pictures? [Y,y,Yes,YES] [N,n,No,NO]"
        answer = gets.chomp!
        if YES.include?(answer) || NO.include?(answer)
          if YES.include?(answer)
            puts "--- YOUR BUG ---"
            print_bug YOU
            puts "\n\n--- MY BUG ---"
            print_bug I
          end
          break
        end
      end
    end
  end

  private
    def play player
      number = Random.rand(1..6)
      case number
      when 1
        if player[:body].eql? 0
          player[:body] = 1
          puts "#{player[:name]} have acquired a body"
        else
          puts "#{player[:name]} already have a body"
        end
      when 2
        one_part player, :neck, :body
      when 3
        one_part player, :head, :neck
      when 4
        two_parts player, :antennaes, :head
      when 5
        two_parts player, :legs, :body
      when 6
        two_parts player, :arms, :body
      end
    end

    def one_part player, part, part_needed
      if player[part].eql? 0
        if player[part_needed].eql? 0
          puts "#{player[:name]} need to have a #{part_needed.to_s} first"
        else
          player[part] = 1
          puts "#{player[:name]} have acquired a #{part.to_s}"
        end
      else
        puts "#{player[:name]} already have a #{part.to_s}"
      end
    end

    def two_parts player, part, part_needed
      if player[part].eql? 0
        if player[part_needed].eql? 0
          puts "#{player[:name]} need to have a #{part_needed.to_s} first"
        else
          player[part] = 1
          puts "#{player[:name]} have acquired first #{part.to_s.chop}"
        end
      else
        if player[part].eql? 2
          puts "#{player[:name]} already have 2 #{part.to_s}"
        else
          player[part] = 2
          puts "#{player[:name]} have acquired second #{part.to_s.chop}"
        end
      end
    end

    def is_completed? player
      player[:body].eql?(1) && player[:neck].eql?(1) && player[:head].eql?(1) && player[:antennaes].eql?(2) && player[:legs].eql?(2) && player[:arms].eql?(2)
    end

    def print_bug player
      antennae_and_leg player, :antennaes
      head player
      neck player
      body player
      antennae_and_leg player, :legs
    end

    def instructions
      loop do
        puts "Do you want an instruction? [Y,y,Yes,YES] [N,n,No,NO]"
        answer = gets.chomp!

        if YES.include?(answer) || NO.include?(answer)
          if YES.include?(answer)
            puts "THE OBJECT OF BUG IS TO FINISH YOUR BUG BEFORE I FINISH"
            puts "MINE. EACH NUMBER STANDS FOR A PART OF THE BUG BODY."
            puts "I WILL ROLL THE DIE FOR YOU, TELL YOU WHAT I ROLLED FOR YOU"
            puts "WHAT THE NUMBER STANDS FOR, AND IF YOU CAN GET THE PART."
            puts "IF YOU CAN GET THE PART I WILL GIVE IT TO YOU."
            puts "THE SAME WILL HAPPEN ON MY TURN."
            puts "IF THERE IS A CHANGE IN EITHER BUG I WILL GIVE YOU THE"
            puts "OPTION OF SEEING THE PICTURES OF THE BUGS."
            puts "THE NUMBERS STAND FOR PARTS AS FOLLOWS:\n\n\n"
            puts "Number    Part          Required Part #"
            puts "1         Body          1"
            puts "2         Neck          1"
            puts "3         Head          2"
            puts "4      (2)antennaes     3"
            puts "5      (2)arms          1"
            puts "6      (2)Legs          1"
          end
          break
        end
      end
    end

    def antennae_and_leg player, part
      if !player[part].eql? 0
        for i in (1...5) do
          if player[part].eql? 1
            puts "     N"
          else
            puts "     N    N"
          end
        end
      end
    end

    def head player
      if !player[:head].eql? 0
        puts "     NNNNNNN"
        puts "     N     N"
        puts "     N O O N"
        puts "     N     N"
        puts "     N  V  N"
        puts "     NNNNNNN"
      end
    end

    def neck player
      if !player[:neck].eql? 0
        puts "     N  N"
      end
    end

    def body player
      if !player[:body].eql? 0
        puts "     NNNNNNNNN"
        for i in (1...5) do
          if i.eql? 2
            if player[:arms].eql? 1
              puts "NNNNNN       N"
            elsif player[:arms].eql? 2
              
              puts "NNNNNN       NNNNNN"
            end
          else
            puts "     N       N"
          end
        end
        puts "     NNNNNNNNN"
      end
    end
end


if __FILE__ == $0
  bug = BugGame.new
  puts "\n\nNOW WE START THE GAME\n\n"
  bug.run
end